query_optimizer/ExecutionHeuristics.hpp

```diff
@@ -0,0 +1,161 @@
+ *   See the License for the specific language governing permissions and
+ *   limitations under the License.
+ **/
+
+#ifndef QUICKSTEP_QUERY_OPTIMIZER_EXECUTION_HEURISTICS_HPP_
+#define QUICKSTEP_QUERY_OPTIMIZER_EXECUTION_HEURISTICS_HPP_
+
+#include <vector>
+
+#include "catalog/CatalogRelation.hpp"
+#include "catalog/CatalogTypedefs.hpp"
+#include "query_execution/QueryContext.pb.h"
+#include "query_optimizer/QueryPlan.hpp"
+#include "utility/Macros.hpp"
+
+#include "glog/logging.h"
+
+namespace quickstep {
+
+namespace optimizer {
+
+/** \addtogroup QueryOptimizer
+ *  @{
+ */
+
+namespace execution_heuristics_internal {
+
+/**
+ * @brief A simple internal class that holds information about various
+ *        hash joins within the execution plan for a query.
+ **/
+class HashJoinInfo {
+ public:
+  HashJoinInfo(QueryPlan::DAGNodeIndex build_operator_index,
+               QueryPlan::DAGNodeIndex join_operator_index,
+               const CatalogRelation *referenced_stored_build_relation,
+               const CatalogRelation *referenced_stored_probe_relation,
+               std::vector<attribute_id> &&build_attributes,
+               std::vector<attribute_id> &&probe_attributes,
+               const QueryContext::join_hash_table_id &join_hash_table_id)
+    : build_operator_index_(build_operator_index),
+      join_operator_index_(join_operator_index),
+      referenced_stored_build_relation_(referenced_stored_build_relation),
+      referenced_stored_probe_relation_(referenced_stored_probe_relation),
+      build_attributes_(build_attributes),
+      probe_attributes_(probe_attributes),
+      join_hash_table_id_(join_hash_table_id) {
+  }
+
+  const QueryPlan::DAGNodeIndex build_operator_index_;
+  const QueryPlan::DAGNodeIndex join_operator_index_;
+  const CatalogRelation *referenced_stored_build_relation_;
+  const CatalogRelation *referenced_stored_probe_relation_;
+  const std::vector<attribute_id> build_attributes_;
+  const std::vector<attribute_id> probe_attributes_;
+  const QueryContext::join_hash_table_id join_hash_table_id_;
+};
+
+}  // namespace execution_heuristics_internal
+
+typedef execution_heuristics_internal::HashJoinInfo HashJoinInfo;
+
+/**
+ * @brief Compiles certain heuristics as an execution plan is being
+ *        converted to a physical plan and uses them to optimize
+ *        the execution plan after it has been generated.
+ **/
+class ExecutionHeuristics {
+ public:
+  static const std::size_t kOneHundred = 100;
+  static const std::size_t kOneThousand = 1000;
+  static const std::size_t kTenThousand = 10000;
+  static const std::size_t kHundredThousand = 100000;
+  static const std::size_t kMillion = 1000000;
+
+  static const std::size_t kCompressionFactor = 10;
+
+  static const std::size_t kVeryLowSparsityHash = 1;
+  static const std::size_t kLowSparsityHash = 2;
+  static const std::size_t kMediumSparsityHash = 5;
+  static const std::size_t kHighSparsityHash = 10;
```

Comment:
I was wondering if we could set above five constants from `kCompressionFactor` to `kHighSparsityHash` as `gflag` variables with the same default values as you set, so that we could benchmark various combinations without recompiling the code.